### ISSUE TITLE
Non standard behaviour converting array to bool

### ISSUE THREAD

**Issue by mjr129** [numpy/numpy#8622]
### Description

Normally, to test if an list, array, dictionary, etc. is None or empty in Python I can use:

```python
if not my_list:
    return "The list is empty!"
else:
    return "The list is not empty!"
```

However, Numpy seems to base this on the lists' content rather than length. Furthermore, it raises an exception if the length is `> 1`. This:

* Is confusing
* Breaks Pythons duck-typing, forcing an expensive recast before I provide Numpy arrays to standard Python functions.
* Doesn't appear to have a strong use case: the current functionality can be reproduced with `a.any` or `a.all` with less ambiguity.

### Steps to reproduce

```python
import numpy

bool( [] )    # False
bool( [0] )    # True
bool( [0, 0] )    # True

bool( {} )    # False
bool( {0: 0} )    # True
bool( {0: 0, 1: 0} )    # True

bool( numpy.array( [] ) ) # False
bool( numpy.array( [0] ) ) # False (???)
bool( numpy.array( [0, 1] ) ) # ValueError: The truth value of an array with more than one element is ambiguous.
```

**Comment by eric-wieser** [comment]
Arguments for these cases:

* `bool( np.array([]) ) == False` - this is consistent with `list` anyway
* `bool( np.array(0) ) == False` - not one of your cases, but this isn't really a list anyway, as you can't call `len` on it. Treating a 0d array as if it were a scalar is reasonable
* `bool( np.array([0, 1]) ) -> ValueError` This is by design, because otherwise code like `if np.array([0, 1]) == 1` would fail silently and always run the following block of code. By making this error, the user is forced to think about `all` vs `any`
* `bool( numpy.array( [0] ) )` - given that we already broke list compatibility above, here it's just easiest to note the array has `.size == 1`, so `all` and `any` are equivalent, and let the user get away with it
* `bool( np.array([[[[0]]]]) )` - as above

**Comment by njsmith** [comment]
Numpy arrays are very much not duck-typed to match lists. 

It is confusing that empty arrays are false-y... it would make more sense for that to raise an error IMO. I can't think of any sensible uses for it either. But probably someone somewhere is depending on it anyway, so it would require a deprecation cycle.

**Comment by eric-wieser** [comment]
@njsmith: I think I'd be in favor of all of the original examples raising `ValueError` in some later release

**Comment by njsmith** [comment]
@eric-wieser: single-element arrays coercing to bool is very widely used, and I think it makes a fair amount of sense. Essentially we're meeting two conflicting constraints: (1) operations on arrays should be elementwise, (2) `__bool__` can only return a single Python scalar (and even if it could return an array, `if` obviously can't do anything sensible with an array; `if array: ...` is by far the most common case where `ndarray.__bool__` gets called). So the intersection of these constraints is: coerce to bool elementwise, when there is only a single element :-)

**Comment by njsmith** [comment]
(I can see an argument from purity that this should only happen for 0d arrays; it's a bit ugly and inconsistent with other situations that we're willing to discard shape information here. But this is arguably a practicality beats purity case, and anyway the disruption of trying to change it is *wayy* not worth it just for a bit of purity.)

**Comment by eric-wieser** [comment]
@njsmith: My argument against single-element arrays is that the existance of that dimension is a strong indication that it might have two-elements next tmie you check it, and so just masks a bug

**Comment by njsmith** [comment]
@eric-wieser: well, that was my intuition, but I admit I haven't actually checked :-). If you wanted to pursue this I guess the first step would be to gather some data... (E.g., by instrumenting numpy and then running various projects' testsuites.)

**Comment by eric-wieser** [comment]
I mean, I'm not saying it's worth the effort necessarily :). But I agree, that would be the best way to proceed.

**Comment by seberg** [comment]
And I always think 0d arrays -> bool makes sense, but single element is funny ;)

**Comment by eric-wieser** [comment]
After #9583 (1.14-dev), `bool( numpy.array( [] ) )` now gives a warning, with the intent to make it error in future.

**Comment by eric-wieser** [comment]
Closing in favor of a summary of a more concise summary in #9885, since AFAICT the only remaining issue here is size-1 but not shape-() arrays.
